torch/csrc/utils/six.h

```diff
@@ -0,0 +1,21 @@
+#pragma once
+
+#include <pybind11/pybind11.h>
+
+namespace six {
+
+// Usually instances of PyStructSequence is also an instance of tuple
+// but in some py2 environment it is not, so we have to manually check
+// the name of the type to determine if it is a namedtupled returned
+// by a pytorch operator.
+
+inline bool isTuple(pybind11::handle input) {
+  std::string m = pybind11::str(input.get_type().attr("__module__"));
+  return pybind11::isinstance<pybind11::tuple>(input) || m == "torch.return_types";
```

Comment:
Uh, this is really expensive!!! In most cases those will be simply tuples, so we should just use `PyTuple_Check` (which is heavily optimized and much faster than regular `isinstance` calls), and only then try to fall back to checking the module like this!

Reply:
More importantly, this PR doesn't refactor all places that use `PyTuple_Check` so we current have a very limited coverage, and will start rejecting things as tuples. What versions of Python fail the `PyTuple_Check` predicate with struct sequences?

Reply:
@apaszke All I know is it fails on some 2.7, but not all 2.7. It seems that its hard to find a pattern which version fails, I even doubt that even for the same version 2.7.9, some build success while other fails. I suggest to look at the build redish of `e3d03bf`.
![image](https://user-images.githubusercontent.com/1032377/52129194-f09c2380-2604-11e9-8611-c70ecdd552c2.png)

Reply:
@apaszke This is a very subtle compatibility issue, I don't think we can replace all `PyTuple_Check` with `six::isTuple`. For example when working on this, I actually tried in https://github.com/pytorch/pytorch/pull/15429/commits/8dfcc4ca6c72e0ecb58fa25e75cbf60d8286a36d to make `python_arg_parser.cpp` support namedtuple to allow something like `ret1 = torch.max(a, dim=0, out=my_namedtuple)`, but his gives segfault. So finally I decided to reverted this change in https://github.com/pytorch/pytorch/pull/15429/commits/304cd44d21e112f446e6fdd2704a5b8eaaba405e and simply not supporting this feature in python2.

Reply:
It's very hard to reproduce this error because it only fails on some build of python2, I can not reproduce it on the python2 on my machine, so all I can do when debugging was to add some print to the code, see the printed message on CI, and decide what to change....

Reply:
No, it looks like all of Python 2 CI is red. Build jobs are green, test jobs are red. And all Caffe2 jobs are obviously green as well. So it seems to be a problem universal in Python 2.

Reply:
What worries me very much is why do we consider those types tuples in there but not elsewhere?! It seems like a very very arbitrary distinction and it will likely break in very weird ways.

Reply:
@apaszke The problem of rejecting `PyStructSequence` as tuple on python 2.7 was not introduced by this PR, but already there since the beginning. So if the user create their own C binding of python using `PyStructSequence`, they would get problem of getting rejected when they pass it as arguments where tuple is required. So the only thing that this PR make it worse is,
it makes much much easier for user to create a `PyStructSequence`, because some pytorch ops natively return `PyStructSequence` object. 

In my opinion, the biggest problem that this PR introduce is:
```python
ret = a.max(dim=0)
if isinstance(ret, tuple):
  do something
```
The above user code would definitely break on python 2.7, even if we have our own `six::isTuple` of `six.is_tuple`, the user would not use it, this is a bc-breaking change...

But since the problem is in CPython, do you think, maybe we should disable this feature completely on python 2.7? Otherwise there would be lots of dirty workaround in pytorch for this issue.